homeassistant/components/nut/__init__.py

```diff
@@ -1 +1,159 @@
 """The nut component."""
+import asyncio
+import logging
+
+from pynut2.nut2 import PyNUTClient, PyNUTError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_ALIAS,
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_RESOURCES,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import DOMAIN, PLATFORMS, PYNUT_DATA, PYNUT_STATUS, PYNUT_UNIQUE_ID
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Network UPS Tools (NUT) component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Network UPS Tools (NUT) from a config entry."""
+
+    config = entry.data
+    host = config[CONF_HOST]
+    port = config[CONF_PORT]
+
+    alias = config.get(CONF_ALIAS)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+
+    data = PyNUTData(host, port, alias, username, password)
+
+    status = await hass.async_add_executor_job(pynutdata_status, data)
+
+    if not status:
+        _LOGGER.error("NUT Sensor has no data, unable to set up")
+        raise ConfigEntryNotReady
+
+    _LOGGER.debug("NUT Sensors Available: %s", status)
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        PYNUT_DATA: data,
+        PYNUT_STATUS: status,
+        PYNUT_UNIQUE_ID: unique_id_from_status(status, host, port),
+    }
+
+    entry.add_update_listener(_async_update_listener)
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+def unique_id_from_status(status, host, port):
+    """Find the best unique id from the status."""
+    # The serial number should always exist, however some off brand
+    # UPSs do not send one so we use the next best thing
+    return status.get("device.serial") or status.get("ups.serial") or f"{host}:{port}"
```

Comment:
We should not use host and port as unique_id as those may change.

Reply:
Is the best way to handle this to turn off any unique ID checks and return None for device_info and unique_id for devices that are missing a serial number?

Reply:
For the config entry it might be easiest to not use a unique_id and just keep track of host and port to make sure there's only one entry per host.

For the entity unique_id we can try to look up the serial and return `None` if the serial is missing.

Reply:
Will abort on the same host/port/alias for the config flow

Will use a combination of Mfr/device/serial as the unique is if a serial is returned (in case some manufacturer reuses serials between product lines) and rerun None if no serial

Reply:
I've made the adjustments